src/elf/symver.rs

```diff
@@ -0,0 +1,907 @@
+            }))
+        }
+
+        /// Get an iterator over the [`Verdef`] entries.
+        #[inline]
+        pub fn iter(&'a self) -> VerdefIter<'a> {
+            self.into_iter()
+        }
+    }
+
+    impl<'a> IntoIterator for &'_ VerdefSection<'a> {
+        type Item = <VerdefIter<'a> as Iterator>::Item;
+        type IntoIter = VerdefIter<'a>;
+
+        #[inline]
+        fn into_iter(self) -> Self::IntoIter {
+            VerdefIter {
+                bytes: self.bytes,
+                count: self.count,
+                index: 0,
+                offset: 0,
+                ctx: self.ctx,
+            }
+        }
+    }
+
+    /// Iterator over the [`Verdef`] entries from the [`SHT_GNU_VERDEF`] section.
+    pub struct VerdefIter<'a> {
+        bytes: &'a [u8],
+        count: usize,
+        index: usize,
+        offset: usize,
+        ctx: container::Ctx,
+    }
+
+    impl<'a> Iterator for VerdefIter<'a> {
+        type Item = Verdef<'a>;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            if self.index >= self.count {
+                None
+            } else {
+                self.index += 1;
+
+                let do_next = |iter: &mut Self| {
+                    let ElfVerdef {
+                        vd_version,
+                        vd_flags,
+                        vd_ndx,
+                        vd_cnt,
+                        vd_hash,
+                        vd_aux,
+                        vd_next,
+                    } = iter.bytes.pread_with(iter.offset, iter.ctx.le).ok()?;
+
+                    // Validate offset to first ElfVerdaux entry.
+                    let offset = iter.offset.checked_add(vd_aux as usize)?;
+
+                    // Validate if offset is valid index into bytes slice.
+                    if offset >= iter.bytes.len() {
+                        return None;
+                    }
+
+                    // Get a slice of bytes starting with the first ElfVerdaux entry.
+                    let bytes: &'a [u8] = &iter.bytes[offset..];
+
+                    // Bump the offset to the next ElfVerdef entry.
+                    iter.offset = iter.offset.checked_add(vd_next as usize)?;
+
+                    Some(Verdef {
+                        vd_version,
+                        vd_flags,
+                        vd_ndx,
+                        vd_cnt,
+                        vd_hash,
+                        vd_aux,
+                        vd_next,
+                        bytes,
+                        ctx: iter.ctx,
+                    })
+                };
+
+                do_next(self).or_else(|| {
+                    // Adjust current index to count in case of an error.
+                    self.index = self.count;
+                    None
+                })
+            }
+        }
+
+        #[inline]
+        fn size_hint(&self) -> (usize, Option<usize>) {
+            let len = self.count - self.index;
+            (len, Some(len))
```

Comment:
```suggestion
            (0, Some(len))
```
We can return less if `sh_info` was wrong, or if the data is corrupt.

Reply:
Good idea.